### PR TITLE
2603

### PR DESCRIPTION
**Pull request by Metalios**
Lots of code fixes to improve performance and reduce potential issues, as well as a few DB updates.

### CODE REVIEW

File: Base/Base/Data_Classes/Archetype.cs [hunk 1]
```diff
@@ -212,33 +201,33 @@ public int CompareTo(object obj)
...
             PopUp.PopupData popupData = new PopUp.PopupData();
-            int index1 = popupData.Add(null);
-            popupData.Sections[index1].Add(this.DisplayName, PopUp.Colors.Title, 1.25f, FontStyle.Bold, 0);
-            popupData.Sections[index1].Add(this.DescShort, PopUp.Colors.Effect, 1f, FontStyle.Bold, 0);
-            popupData.Sections[index1].Add(this.DescLong, PopUp.Colors.Text, 0.9f, FontStyle.Bold, 1);
-            int index2 = popupData.Add(null);
-            popupData.Sections[index2].Add("You can't change archetype once a build has been started.\nIf you want to pick a different archetype, you need to clear the current build and start a new one.", PopUp.Colors.Effect, 0.9f, FontStyle.Bold, 0);
+            int index1 = popupData.Add();
+            popupData.Sections[index1].Add(DisplayName, PopUp.Colors.Title, 1.25f);
+            popupData.Sections[index1].Add(DescShort, PopUp.Colors.Effect);
```
Comment (ImaginaryDevelopment) [inline Base/Base/Data_Classes/Archetype.cs:227]: This appears to be a change in substance, accident or on purpose? Bold is lost.

File: Base/Files.cs [hunk 2]
```diff
@@ -33,19 +34,17 @@ string SearchUpRec(string foldername, string filename)
             // get the directory that holds the filename, filename should be a FULL path
             var fnDir = Path.GetDirectoryName(filename);
             // if the filename is already in a folder with the correct foldername, we need to go up twice instead of once.
-            var targetRoot = fnDir.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar).EndsWith(foldername) ?
+            var targetRoot = fnDir != null && fnDir.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar).EndsWith(foldername) ?
                     Path.GetDirectoryName(fnDir) : fnDir;
             var parent = Path.GetDirectoryName(targetRoot);
             if (parent == null) return null;
-            var attempt = Path.Combine(Path.Combine(parent, foldername), Path.GetFileName(filename));
+            var attempt = Path.Combine(Path.Combine(parent, foldername), Path.GetFileName(filename) ?? throw new InvalidOperationException());
```
Comment (ImaginaryDevelopment) [inline Base/Files.cs:41]: Path.Combine throws on either one being null, so your `??` will never fire. see: https://referencesource.microsoft.com/#mscorlib/system/io/path.cs,090eca8621a248ee,references

File: Base/Files.cs [hunk 3]
```diff
@@ -33,19 +34,17 @@ string SearchUpRec(string foldername, string filename)
             // get the directory that holds the filename, filename should be a FULL path
             var fnDir = Path.GetDirectoryName(filename);
             // if the filename is already in a folder with the correct foldername, we need to go up twice instead of once.
-            var targetRoot = fnDir.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar).EndsWith(foldername) ?
+            var targetRoot = fnDir != null && fnDir.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar).EndsWith(foldername) ?
```
Comment (ImaginaryDevelopment) [inline Base/Files.cs:37]: why not use `?.TrimEnd` ?

File: Base/Modifiers.cs [hunk 4]
```diff
@@ -115,31 +111,28 @@ public bool Load()
                 fileStream.Close();
                 return false;
             }
-            else
+
+            Revision = reader.ReadInt32();
```
Comment (ImaginaryDevelopment) [inline Base/Modifiers.cs:115]: This was being done inside a loop before, now that it isn't the stream won't be moved forward properly.

File: Base/MidsCharacterFileFormat.cs [hunk 5]
```diff
@@ -67,13 +67,13 @@ static bool MxDBuildSaveBuffer(ref byte[] buffer, bool includeAltEnh)
                 writer.Write(Convert.ToBoolean(power.StatInclude));
                 writer.Write(power.VariableValue);
                 writer.Write(Convert.ToSByte(power.SubPowers.Length - 1));
-                for (int index2 = 0; index2 < power.SubPowers.Length; ++index2)
+                foreach (var index2 in power.SubPowers)
```
Comment (ImaginaryDevelopment) [inline Base/MidsCharacterFileFormat.cs:70]: this looks risky. it was using a counter, now you are using the array values. If the array values are just a count, then no harm, but I'm not sure that's the case, are you?

File: Hero Designer/Data/db/Other/_Mastermind_Pets.Ember_Demon.json [hunk 6]
```diff
@@ -367,177 +367,7 @@
       "Ignore_Buff": [],
       "MutexAuto": true,
       "ForcedClass": "Class_Minion_Henchman",
-      "Effects": [
```
Comment (ImaginaryDevelopment) [inline Hero Designer/Data/db/Other/_Mastermind_Pets.Ember_Demon.json:370]: was all this removed intentionally?

**Review by ImaginaryDevelopment — COMMENTED** [review]
check out the comments

File: Base/Base/Data_Classes/Archetype.cs (hunk 1 above)
Comment (Metalios) [reply Base/Base/Data_Classes/Archetype.cs:227]: Was an accidental change, although I did compare 2.6.0.2's popup info vs these changes and there doesn't appear to be any downside to this.

File: Base/Files.cs (hunk 3 above)
Comment (Metalios) [reply Base/Files.cs:37]: Makes sense, honestly as of right now I do not know why I did this that way.

File: Base/Modifiers.cs (hunk 4 above)
Comment (Metalios) [reply Base/Modifiers.cs:115]: This is one I have reverted but apparently I didn't push it. This is what it should be..

````
public bool Load()
    {
        string path = Files.SelectDataFileLoad("AttribMod.mhd");
        Modifier = new ModifierTable[0];
        FileStream fileStream;
        BinaryReader reader;
        try
        {
            fileStream = new FileStream(path, FileMode.Open, FileAccess.Read);
            reader = new BinaryReader(fileStream);
        }
        catch (Exception ex)
        {
            MessageBox.Show(ex.Message + '\n' + '\n' + "Modifier tables couldn't be loaded.");
            return false;
        }
        try
        {
            if (reader.ReadString() != "Mids' Hero Designer Attribute Modifier Tables")
            {
                MessageBox.Show("Modifier table header wasn't found, file may be corrupt!");
                reader.Close();
                fileStream.Close();
                return false;
            }

            Revision = reader.ReadInt32();
            RevisionDate = DateTime.FromBinary(reader.ReadInt64());
            SourceIndex = reader.ReadString();
            SourceTables = reader.ReadString();
            int num = 0;
            Modifier = new ModifierTable[reader.ReadInt32() + 1];
            for (int index = 0; index <= Modifier.Length - 1; ++index)
            {
                Modifier[index] = new ModifierTable();
                Modifier[index].Load(reader);
                if (num <= 5)
                    continue;
                num = 0;
                Application.DoEvents();
            }
            return true;
        }
        catch (Exception ex)
        {
            MessageBox.Show("Modifier table file isn't how it should be (" + ex.Message + ")" + '\n' + "No modifiers loaded.");
            Modifier = new ModifierTable[0];
            reader.Close();
            fileStream.Close();
            return false;
        }
    }
````

File: Base/MidsCharacterFileFormat.cs (hunk 5 above)
Comment (Metalios) [reply Base/MidsCharacterFileFormat.cs:70]: I'm pretty sure, I have tested the loading of builds from various versions after making these changes and all load as expected.

File: Base/Modifiers.cs (hunk 4 above)
Comment (ImaginaryDevelopment) [reply Base/Modifiers.cs:115]: I could have been reading the diff wrong.

File: Hero Designer/Data/db/Other/_Mastermind_Pets.Ember_Demon.json (hunk 6 above)
Comment (Metalios) [reply Hero Designer/Data/db/Other/_Mastermind_Pets.Ember_Demon.json:370]: Yes and a resounding no at the same time. This is the portion where somehow I have lost my notes and either need to A) Track down how I implemented the changes to the 2 MM pets or B) Remove it and redo it from scratch.

File: Base/Files.cs (hunk 2 above)
Comment (Metalios) [reply Base/Files.cs:41]: I see it

**Pull request closed without merging**